From 912d8e5caae85c7d7bf3f2ee8b74aca9f68ac8a3 Mon Sep 17 00:00:00 2001
From: Fugang Duan <b38611@freescale.com>
Date: Wed, 5 Aug 2015 18:21:04 +0800
Subject: [PATCH 0222/5242] MLK-11325 dma: mxs-dma: fix dma_io clock enable
 count mismatch issue

commit  ec409cfa80ba4fc2edbc31b124608a54116e5784 from
https://source.codeaurora.org/external/imx/linux-imx.git

After dma init by calling .mxs_dma_init(), disable dma_io and
dma_bch clocks. When dma chans are requested by devices, clocks
are enabled in .device_alloc_chan_resources(). The patch is to
fix clock enable count mismatch issue.

Signed-off-by: Fugang Duan <B38611@freescale.com>
(cherry picked from commit: 4868cf5e39a0aeb1ad12c5c1a453d233c0f472ce)
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/dma/mxs-dma.c |    9 ++++++---
 1 file changed, 6 insertions(+), 3 deletions(-)

diff --git a/drivers/dma/mxs-dma.c b/drivers/dma/mxs-dma.c
index 5913abf..d5e7c2a 100644
--- a/drivers/dma/mxs-dma.c
+++ b/drivers/dma/mxs-dma.c
@@ -717,12 +717,12 @@ static int mxs_dma_init(struct mxs_dma_engine *mxs_dma)
 	if (mxs_dma->dev_id == IMX7D_DMA) {
 		ret = clk_prepare_enable(mxs_dma->clk_io);
 		if (ret)
-			goto err_out;
+			goto err_clk_bch;
 	}
 
 	ret = stmp_reset_block(mxs_dma->base);
 	if (ret)
-		goto err_out;
+		goto err_clk_io;
 
 	/* enable apbh burst */
 	if (dma_is_apbh(mxs_dma)) {
@@ -736,7 +736,10 @@ static int mxs_dma_init(struct mxs_dma_engine *mxs_dma)
 	writel(MXS_DMA_CHANNELS_MASK << MXS_DMA_CHANNELS,
 		mxs_dma->base + HW_APBHX_CTRL1 + STMP_OFFSET_REG_SET);
 
-err_out:
+err_clk_io:
+	if (mxs_dma->dev_id == IMX7D_DMA)
+		clk_disable_unprepare(mxs_dma->clk_io);
+err_clk_bch:
 	clk_disable_unprepare(mxs_dma->clk);
 	return ret;
 }
-- 
1.7.9.5

